src/base/Notifier.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import {ISubscriber} from "../interfaces/ISubscriber.sol";
+import {PositionConfig} from "../libraries/PositionConfig.sol";
+
+contract Notifier {
+    error SubscriberCannotBeNotified();
+    error AlreadySubscribed(address subscriber);
+
+    event Subscribed(uint256 tokenId, address subscriber);
+    event Unsubscribed(uint256 tokenId, address subscriber);
+
+    uint256 private immutable subscriberGasLimit;
+    ISubscriber private constant NO_SUBSCRIBER = ISubscriber(address(0));
+
+    mapping(uint256 tokenId => ISubscriber subscriber) public subscriber;
+
+    constructor(uint256 _subscriberGasLimit) {
+        subscriberGasLimit = _subscriberGasLimit;
+    }
+
+    function _subscribe(uint256 tokenId, PositionConfig memory config, address newSubscriber) internal {
+        ISubscriber _subscriber = subscriber[tokenId];
+
+        if (_subscriber != NO_SUBSCRIBER) revert AlreadySubscribed(address(_subscriber));
+        subscriber[tokenId] = ISubscriber(newSubscriber);
+
+        ISubscriber(newSubscriber).notifySubscribe(tokenId, config);
+        emit Subscribed(tokenId, address(newSubscriber));
+    }
+
+    /// @dev Must always allow a user to unsubscribe. In the case of a malicious subscriber, a user can always unsubscribe safely, ensuring liquidity is always modifiable.
+    function _unsubscribe(uint256 tokenId, PositionConfig memory config) internal {
+        ISubscriber _subscriber = subscriber[tokenId];
+
+        if (gasleft() < subscriberGasLimit) revert SubscriberCannotBeNotified();
```

Comment:
i feel like this revert is unnecessary - maybe this subscriber only needs like 30% of the total subscriberGasLimit, but the user has to send WAY more just to get this line to pass. If it actually runs out of gas then that will cause the failure?

Reply:
i feel like this is true in core too... like what does that check gain

Reply:
i think youre right, what I thought it was preventing was a user sending too little gas to evade the notification to subscriber but I think it would still just revert later

Reply:
yeah same... i feel like its wrong in core too

Reply:
Let me open an issue in core as well then?

Reply:
https://github.com/Uniswap/v4-core/issues/814